core/src/connection_reuse.rs

```diff
@@ -40,205 +40,556 @@
+                "Clearing existing connection to {} which errored earlier: {:?}",
+                addr,
+                err
+            );
+        } else {
+            return false;
+        }
+
+        // only reaching the point if the entry was an error
+        conns.remove(addr);
+        return true;
+    }
+}
+
+/// Polls the given outbound of addr for a new substream. Returns `Ok(Some(stream))` if available,
+/// `Ok(None)` if the stream is not ready yet add an `Err` if the stream had failed.
+fn poll_for_substream<O, S>(mut outbound: O, addr: &Multiaddr)
+    -> Result<Option<S>, IoError>
+where
+
+    S: AsyncRead + AsyncWrite,
+    O: Future<Item = Option<S>, Error = IoError>
+{
+    match outbound.poll() {
+        Ok(Async::Ready(Some(inner))) => {
+            trace!("Opened new outgoing substream to {}", addr);
+            // all good, return the new stream
+            Ok(Some(inner))
+        }
+        Ok(Async::NotReady) => Ok(None),
+        Ok(Async::Ready(None)) => {
+            // The muxer can no longer produce outgoing substreams.
+            // Let's reopen a connection.
+            trace!("Closing existing connection to {} ; can't produce outgoing substreams", addr);
+
+            let err = IoError::new(IoErrorKind::ConnectionRefused, "No Streams left");
+            Err(err)
+        }
+        Err(err) => {
+            // If we get an error while opening a substream, we decide to ignore it
+            // and open a new muxer.
+            // If opening the muxer produces an error, *then* we will return it.
+            debug!(
+                "Error while opening outgoing substream to {}: {:?}",
+                addr, err
+            );
+            Err(err)
+        }
+    }
+}
+
+impl<T, C> ConnectionsManager<T, C>
+where
+    T: Transport,
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture> + Clone,
+    C::Output: StreamMuxer + 'static,
+    UpgradedNode<T, C>: Transport<Output = C::Output> + Clone,
+    <UpgradedNode<T, C> as Transport>::Dial: 'static,
+    <UpgradedNode<T, C> as Transport>::MultiaddrFuture: 'static,
+{
+    /// Informs the Manager about a new inbound connection that has been
+    /// established, so it
+    fn new_inbound(&self, addr: &Multiaddr, listener_id: usize, muxer: C::Output) {
+        let mut conns = self.connections.lock();
+        let new_state = PeerState::Active {
+            muxer: Arc::new(muxer),
+            listener_id: Some(listener_id),
+            closed: Arc::new(AtomicBool::new(false))
+        };
+        let mut old_state = conns.insert(addr.clone(), new_state);
+
+        if let Some(PeerState::Pending { ref mut notify, .. }) = old_state {
+            // we need to wake some up, that we have a connection now
+            trace!("Found incoming for pending connection to {}", addr);
+            for (_, task) in notify.drain() {
+                task.notify()
+            }
+        } else {
+            old_state.map(|s| s.reset_conn());
```

Comment:
This is very bad style in my opinion. You're mixing a "pulling" API (`map`) with a "pushing" action (`reset_conn`).